src/mpid/ch4/shm/src/topotree.c

```diff
@@ -11,75 +11,134 @@
-                                  bool package_leaders_first);
-
-static int gen_tree(int k_val, int *shared_region, int num_packages,
-                    int **ranks_per_package, int max_ranks_per_package, int *package_ctr,
-                    int package_level, int num_ranks, bool package_leaders_first,
-                    bool right_skewed, MPIR_Errflag_t * eflag);
-
-/* This function allocates and generates a template_tree based on k_val and right_skewed for
- * max_ranks.
+/* This function allocates and generates a template_tree based on k_val, tree_type, and
+ * skewness for max_ranks.
  * */
-static int create_template_tree(MPIDI_SHM_topotree_t * template_tree, int k_val,
-                                bool right_skewed, int max_ranks, MPIR_Errflag_t * errflag)
+int MPIDI_SHM_create_template_tree(MPIDI_SHM_topotree_t * template_tree, int k_val, int tree_type,
+                                   bool right_skewed, int max_ranks, MPIR_Errflag_t * errflag)
 {
     int mpi_errno = MPI_SUCCESS, mpi_errno_ret ATTRIBUTE((unused)) = MPI_SUCCESS;
-    int child_id, child_idx;
+    int i, j, child_id, child_idx;
 
     mpi_errno = MPIDI_SHM_topotree_allocate(template_tree, max_ranks, k_val);
     MPIR_ERR_COLL_CHECKANDCONT(mpi_errno, *errflag);
 
-    for (int i = 0; i < max_ranks; ++i) {
-        MPIDI_SHM_TOPOTREE_PARENT(template_tree, i) = ceilf(i / (float) (k_val)) - 1;
-        MPIDI_SHM_TOPOTREE_NUM_CHILD(template_tree, i) = 0;
-        if (!right_skewed) {
-            for (int j = 0; j < k_val; ++j) {
-                child_id = i * k_val + 1 + j;
-                if (child_id < max_ranks) {
-                    child_idx = MPIDI_SHM_TOPOTREE_NUM_CHILD(template_tree, i)++;
-                    MPIDI_SHM_TOPOTREE_CHILD(template_tree, i, child_idx) = child_id;
+    if (tree_type == MPIDI_POSIX_RELEASE_GATHER_TREE_TYPE_KARY) {
+        /* Notice that skewness matters only for kary trees */
+        for (i = 0; i < max_ranks; ++i) {
+            MPIDI_SHM_TOPOTREE_PARENT(template_tree, i) = ceilf(i / (float) (k_val)) - 1;
+            MPIDI_SHM_TOPOTREE_NUM_CHILD(template_tree, i) = 0;
+            if (!right_skewed) {
+                for (j = 0; j < k_val; ++j) {
+                    child_id = i * k_val + 1 + j;
+                    if (child_id < max_ranks) {
+                        child_idx = MPIDI_SHM_TOPOTREE_NUM_CHILD(template_tree, i)++;
+                        MPIDI_SHM_TOPOTREE_CHILD(template_tree, i, child_idx) = child_id;
+                    }
+                }
+            } else if (right_skewed) {
+                for (j = k_val - 1; j >= 0; --j) {
+                    child_id = i * k_val + 1 + j;
+                    if (child_id < max_ranks) {
+                        child_idx = MPIDI_SHM_TOPOTREE_NUM_CHILD(template_tree, i)++;
+                        MPIDI_SHM_TOPOTREE_CHILD(template_tree, i, child_idx) = child_id;
+                    }
```

Comment:
Extending the tree functions here seems unrelated to the nonblocking framework. There is a lot going on in this commit. From what I can see:
1. nonblocking framework
2. new collectives (ibcast, ireduce)
3. new tree type (knomial1)

It would be helpful for review to separate these features out. This is quite a challenge in its current form.

Reply:
I second @raffenet. This commit (ch4/shm: add a new tree type for shm ) is too messy.